File: go/array/1_two_sum.go
package array

// O(n) Hash Map
func twoSum1(nums []int, target int) []int {

	// hashMap 将一个num 放到hMap 中，查看另一个target - num 是不是也在其中
	// key 是num， val 是idx
	hmap := make(map[int]int, len(nums))

	for i, v := range nums {
		if idx, ok := hmap[target-v]; ok {
			return []int{i, idx}
		} else {
			hmap[v] = i
		}
	}

	return []int{}
}

// O(n2) 暴力遍历
func twoSum2(nums []int, target int) []int {

	for i := 0; i < len(nums); i++ {
		for j := i + 1; j < len(nums); j++ {
			if nums[i]+nums[j] == target {
				return []int{i, j}
			}
		}
	}
	return []int{}
}

File: rust/src/array/1_two_sum.rs
use::std::collections::HashMap;



pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {

    let mut hmap: HashMap<_, usize> = HashMap::with_capacity(nums.len());

   // 取出值具体匹配 2
   for i in 0..nums.len(){
        let val = nums[i];
        if let Some(k) = hmap.get(&(target - val)) {
            if *k != i {
                return vec![i as i32, *k as i32]
            }
        }

        hmap.insert(val, i);
   }

   panic!("not found")


    // 取出值具体匹配 1 leetcode 不知道为什么编译不通过
    // for (i, item) in nums.iter().enumerate(){
    //     if let Some(k) = hmap.get(&(target - *item)) {
    //         // k 能匹配到意味着有值
    //         if *k != i {
    //             return vec![i as i32, *k as i32]
    //         }
    //     }

    //     hmap.insert(item, i);
    // }


    // contains 判断 fail 不能输出idex
    // for (i, item) in nums.iter().enumerate() {
    //     if hmap.contains_key(&(target - *item)) {
    //         return vec![*item, target-*item]
    //     }else {
    //         hmap.insert(item, i);
    //     }
        
    // }

}